onnxruntime/contrib_ops/cpu/transformers/logits_processor.cc

```diff
@@ -0,0 +1,192 @@
+  int batch_beam_size = next_token_scores.batch_beam_size;
+
+  for (int i = 0; i < batch_beam_size; i++) {
+    gsl::span<T> beam_token_scores = next_token_scores.GetScores(i);
+    gsl::span<const int64_t> sequence = sequences->GetSequence(i);
+
+    gsl::span<const int64_t> prefix = sequence.subspan(sequence.length() - prefix_length);
+    ORT_ENFORCE(prefix.length() == prefix_length);
+
+    std::unordered_set<int64_t> blocked_word_ids;
+    for (int j = 0; j <= static_cast<int>(sequence.length()) - ngram_size_; j++) {
+      // Here we use naive algorithm for matching. The complexity is O(batch_beam_size * ngram_size * sequence_length)
+      // TODO: build N-Gram index (hash table with prefix of length NGram - 1 as key, and list of last word of NGram as value) for fast matching.
+      if (ngram_size_ == 1 || prefix == sequence.subspan(j, prefix_length)) {
+        blocked_word_ids.insert(sequence[j + prefix_length]);
+      }
+    }
+
+    for (const int64_t word_id : blocked_word_ids) {
+      beam_token_scores[word_id] = std::numeric_limits<T>::lowest();
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpScores("NoRepeatNGramLogitsProcessor", next_token_scores.scores);
+#endif
+}
+
+template <typename T>
+VocabMaskLogitsProcessor<T>::VocabMaskLogitsProcessor(const gsl::span<const int32_t>& vocab_mask) : vocab_mask_(vocab_mask) {
+}
+
+template <typename T>
+void VocabMaskLogitsProcessor<T>::Process(const ISequences* /*sequences*/,
+                                          NextTokenScores<T>& next_token_scores) {
+  assert(!vocab_mask_.empty());
+
+  // Process vocabulary mask and set tokens with mask value 0 to -inf.
+  T* p = next_token_scores.scores.data();
+  // next_token_scores shape (batch_size * num_beams, vocab_size)
+  // vocab_mask shape (vocab_size). TODO: support shape (batch_size, vocab_size)
+  for (int i = 0; i < next_token_scores.batch_beam_size; i++) {
+    for (int j = 0; j < next_token_scores.vocab_size; j++, p++) {
+      if (vocab_mask_[j] == 0) {
+        *p = std::numeric_limits<T>::lowest();
+      }
+    }
+  }
+
+#ifdef DEBUG_BEAM_SEARCH
+  DumpScores("VocabMaskLogitsProcessor", next_token_scores.scores);
+#endif
+}
+
+template <typename T>
+void LogitsProcessorList<T>::Init(const BeamSearchParameters& parameters) {
+  processor_list_.clear();
+
+  if (parameters.repetition_penalty != 1.0f) {  // 1.0 means no penalty
+    repetition_penalty_processor_ = std::make_unique<RepetitionPenaltyLogitsProcessor<T>>(parameters.repetition_penalty);
+    processor_list_.push_back(repetition_penalty_processor_.get());
+  }
+
+  if (parameters.no_repeat_ngram_size > 0) {
+    no_repeat_ngram_processor_ = std::make_unique<NoRepeatNGramLogitsProcessor<T>>(parameters.no_repeat_ngram_size);
+    processor_list_.push_back(no_repeat_ngram_processor_.get());
+  }
+
+  if (!parameters.vocab_mask.empty()) {
+    vocab_mask_processor_ = std::make_unique<VocabMaskLogitsProcessor<T>>(parameters.vocab_mask);
```

Comment:
vocab mask supplied is used to zero out unrelated vocabs.
Semantically, shouldn't this precede any other processing - repetition or no_repeat?

Reply:
Here, we uses the same order as in huggingface transformer to get parity result:

https://github.com/huggingface/transformers/blob/e513c16e82221e0a150b97a31f2c3c864913c454/src/transformers/generation_utils.py#L610-L639

NoBadWordsLogitsProcessor is after RepetitionPenaltyLogitsProcessor and NoRepeatNGramLogitsProcessor.